File: python/kickstart/2020/2020_c/3_perfect_subarray.py
from collections import Counter
from math import sqrt


def input_numbers(t=int):
    return map(t, input().split())


Sq = [i*i for i in range(int(sqrt(10**7)))]

T, = input_numbers()
for _c in range(1, T+1):
    N, = input_numbers()
    A = list(input_numbers())
    cnt = Counter()
    ans = 0
    pre = 0
    cnt[0] = 1
    mi = 0
    for v in A:
        pre += v
        for d in Sq:
            if pre - d < mi:
                break
            ans += cnt[pre - d]
        mi = min(mi, pre)
        cnt[pre] += 1
    print(f"Case #{_c}: {ans}")

File: kotlin/src/main/kotlin/google_kickstart/2020/C/3.perfect_subarray.kt
package google_kickstart.`2020`.C

import kotlin.math.sqrt

fun main() {
    fun split_input(): List<String> = readLine()!!.trim().split(" ")
    fun input_ints() = split_input().mapNotNull { if (it != "") it.toInt() else null }
    fun input_longs() = split_input().mapNotNull { if (it != "") it.toLong() else null }

    val sq = (0..sqrt(1e8).toInt()).map { it * it }

    val T = readLine()!!.toInt()
    (1..T).forEach {
        val (N) = input_ints()
        val P = input_ints().scan(0, Int::plus)
        val cnt = hashMapOf<Int, Int>()
        cnt[0] = 1

        var (res, min) = 0L to 0
        for (v in P.subList(1, P.size)) {
            for (d in sq) {
                if (v - d < min) break
                res += cnt.getOrDefault(v - d, 0)
            }

            cnt[v] = cnt.getOrDefault(v, 0) + 1
            min = minOf(min, v)
        }
        println("Case #${it}: $res")
    }
}
